### ISSUE TITLE
Make lr_find() usable in a distributed data parallel mode, a  step towards enabling notebook to run distributed training.

### ISSUE THREAD

**Issue by philtrade** [fastai/fastai#2485]
In the process of building a tool to facilitate distributed training + `fastai v1` in Jupyter notebook, I realized `lr_find()` doesn't work in distributed mode, and it doesn't need a distributed learner to work either.

Although I got it to work by dynamically intercepting`Learner.lr_find()`, I have distilled that monkey wrench/patch down to a backward compatible, fairly simple change.  I'd like to submit it here for your consideration.

The net outcome is that: one can use a Learner which had been `to_distributed()`,  and run `lr_find()` on it without any complication.  It is a beneficial feature to fastai's Jupyter notebook users, IMHO.

If it pleases you, I'll be happy to submit a PR, and help porting it over to `fastai v2` as well, so that `fastai v2` lesson notebooks will be *distributed-training ready* early on, a feature that was asked and answered "no" many times for `fastai v1`.

The change takes 2 steps:

1. Only allows the master proc (rank 0) to run `lr_find()`, like how `basic_train.save()` does it, other GPUs aren't helpful anyway to this training run, are they?

2. Temporarily take the `learner` out of the distributed mode, do the usual `lr_find()` dance, and restore it back to the distributed condition afterwards.

Below is the diff, I have tested it *with and without distributed training, with one GPU or multiple GPUs DDP*.  Comments are welcome!  I'll be formalize it into a PR if it's acceptable to you. 

* ["The diff for *Temporarily suspend distributed mode when doing lr_find*"](https://github.com/fastai/fastai/compare/master...philtrade:master?diff=split)

Thank you.

Note: [1]. the reason DDP doesn't play well for `lr_find()` is that, the multiple GPUs each trains on its own and may early-stop at different batches, causing spurious deadlocks, as the earlier quitters and late comers wait for each other at different rendezvous points.  Hence the above solution.
[2]  The tool I built, [The iPython extension `Ddip`](https://github.com/philtrade/Ddip), which I intend to add support to `fastai v2` soon, and would love to contribute it to  the evolving `fastai` code base.

**Comment by sgugger** [comment]
Running lr find in process 0 only will change the effective batch size compared to the distributed training. I think the check should be to let the training run its full course on all processes for this, to avoid the dead-lock, then just showing the loss of process 0 up to the point this process would have stopped.

**Comment by philtrade** [comment]
Thanks for the response Sylvain.  I wanna touch upon your suggested implementation, then share my observations from experiments on the effect of batch size (`bs` when creating the training data) in distributed training.  In short, a single absolute **number of batches** seems to matter, and I'm not sure if that's a known behavior, and if that's the same thing as the "effective batch size" you mentioned here, and that in distributed training with multiple GPU, this number becomes too short, without adjusting the `bs` used to create the training data.

### On the implementation you proposed

> *"run its full course on all processes"*

Do you mean in `LRFinder.on_batch_end()`, limit the early stopping by expanding losses to non-distributed training?  E.g. `(smooth_loss > 4*self.best_loss and not num_distrib())`? 

> *"showing the loss of process 0 up to the point this process would have stopped."*

Something like: let `DistributedRecorder.on_train_end()` do the `smooth_loss > 4 * best_loss` logic, and trim the `recorder.{losses,lrs}` up to the rediscovered *best loss* ?

This sounds like a feasible lock-free approach. 

Nonetheless we remain facing the issue below, when running `lr_find()` in distributed mode.

### On the impact of `bs` to `lr_find()` in distributed training:

From my observation, lr_find needs enough *number of batches* as a runway to discover the run-away losses, which **curiously, seems to be independent of the batch size, and of the number of GPUs used in distributed training**.   And because in distributed mode the *number of batches* is reduced (divided by the number of GPUs) at the same batch size, the runway got shorter, sometimes too short.

Using the DL1, lesson3-camvid as an example, on a GTX1080Ti, typically it needs around 68-72 batches to hit the loss expansion.

#### When `lr_find()` works in non-distributed single GPU mode, with long enough runway:

First three charts shows that in single GPU training, how `lr_find()` works OK with bs=8,4,2, 
andthe total # of batches per epoch is 75, 150, and 300 respectively.  However, `lr_find()` stops at the similar proximity of batch # 70, # 72, and # 68 respectively.

Notice the batch number at which `lr_find()` stopped at, in the output
>> `Rank [0] LR Finder is complete @ batch# ....`

![bs=8, non-DDP](https://user-images.githubusercontent.com/19887230/73792729-c0c4b600-4759-11ea-826b-21ba974af0f5.png)

![bs=4 non-DDP](https://user-images.githubusercontent.com/19887230/73792902-25801080-475a-11ea-9f91-051d850e9a5b.png)

![bs=2 non-DDP](https://user-images.githubusercontent.com/19887230/73792918-2fa20f00-475a-11ea-9367-bd3c3131a093.png)

### But in Distributed Training, Total # of batches is scaled down, and becomes too short for `lr_find()` -- it needs 70-ish !

Now when in DDP mode with 3 GPUs, `lr_find()` exhausts the run without early_stopping at bs=8 and 4, because total *number of batches* per epoch are scaled down from 75 and 150, by 3,  to **25 and 50**.  The `bs=8` run `lr_find()` automatically scales up to 2 epochs, but still, 25x2=50, falls short.  **Notice the shorter learning rate range of the x-axis, when compared to above single GPU mode**.

![bs=8 DDP bn=25, epoch = 2](https://user-images.githubusercontent.com/19887230/73794090-7ee93f00-475c-11ea-988b-7b87e98cf25c.png)

![bs=4 DDP bn=50](https://user-images.githubusercontent.com/19887230/73794162-b0faa100-475c-11ea-83de-7fb78b09100a.png)

**Changing `num_it` doesn't help:**  @ `bs=8`, lr_find() runs with`num_it=200`

![bs=8 DDP bn=25 num_it = 200](https://user-images.githubusercontent.com/19887230/73795123-b9ec7200-475e-11ea-9b8f-f0a3cf4c0836.png)

**Finally `lr_find()` in distributed training mode finds the `best_loss`**, when @ `bs=2` because the runway becomes long enough: 300/3 = 100 (> 72).

![bs=2 DDP bn=100](https://user-images.githubusercontent.com/19887230/73795023-7560d680-475e-11ea-8dd6-539b36b3f215.png)

**But @ `bs=2`, # of batches = 100, it spuriously dead lock**, when one process early-stopped, and waiting for other process at `on_train_end()`'s `learn.load('tmp')` call, **while other processes still churning next batches**, and `on_batch_end()`, they ended up waiting for the early-quitter to sync up to reduce the losses:

![bs=2 lr_find DDP deadlock](https://user-images.githubusercontent.com/19887230/73795488-9a097e00-475f-11ea-8a3d-f63b259b1a0a.png)

####  Apparently, with multiple GPU, effective batch size maybe bigger, but total **number** of batches to step through is smaller.  As the learning rate is stepped in `LRFinder.on_batch_end()`,  with fewer batches, it covers shorter range of learning rate. 

What are the solutions to this? 

When building `Ddip`, I strived for **minimal demand on users to tweak the notebook code**,  and this demand on user to "adjust `bs` just for the `lr_find()` then restore it again for actual training", seems not very user-friendly.

I went to the route of running `lr_find()` in  "temporarily non-distributed mode, on the rank-0 process",  such that users won't be surprised that their prior experimental settings, and output from `lr_find()`, stopped behaving the same in distributed training. 

But I'm definitely open to brain storming on better solutions!!

Sorry about this long posts, it took me a long time to debug the deadlock and understand this much.  Thanks.

**Comment by philtrade** [comment]
Hi Sylvain,

Increasing epochs in distributed `lr_find()` seems to address the issue of not enough number of batches to reach the point of expanding losses:

```
def lr_find(learn:Learner, start_lr:Floats=1e-7, end_lr:Floats=10, num_it:int=100, stop_div:bool=True, wd:float=None):
    ...
    epochs = int(np.ceil(num_it/len(learn.data.train_dl)))
 +  if num_distrib() > 1: epochs *= num_distrib()
    learn.fit(epochs, start_lr, callbacks=[cb], wd=wd)
```
Now `bs=8`, 3 GPUs, `epochs` scaled up from 2 to 6, `lr_find()` would early-stop in the 3rd epoch.  Previously it ran out after 2 epochs without seeing loss expansion.

Next will try the deadlock avoidance approach you mentioned --- let them run the full course.

**Comment by philtrade** [comment]
@sgugger, 

Progress on avoiding the deadlock following your suggestion:

1. In `LRFinder.on_batch_end()`, do not abort the epoch in distributed mode, let it finish.  Defer to aborting the training in `on_epoch_end()`.  Also print a suggestion on `plot(skip_end=..)` value to use.

2. Also scale up the number of epochs to use in `lr_find()` by the number of GPUs , to give `LRFinder` more runway stepping through the range of learning rates adequately.

In my testing, GPUs synced up happily at end of epoch.  Time overhead is only a partial epoch after loss expansion,  despite the extended number of epochs.  I've conducted 100 runs of the `bs=1` case, and no deadlock.

Really appreciate your suggestion.   Could you be so kind to review this short patch?  If it looks acceptable, I will draft a PR.

 https://github.com/fastai/fastai/compare/master...philtrade:master?diff=unified

Thank you.

**Comment by sgugger** [comment]
Fixed by #2487
